### ISSUE TITLE
Use parameters & dynamic reconfigure for setting differential drive constants

### ISSUE THREAD

**Issue by jay3ss** [CSULATeleop/teleop#3]
This is more of an enhancement than an issue. Currently, the constants used to scale the `geometry_msgs/Twist` messages received by the differential driver (`LEFT_SCALE` and `RIGHT_SCALE`) are hardcoded. This makes it tedious to change the constants as each change requires recompilation and uploading to the Arduino. 

By using [parameters](http://wiki.ros.org/Parameter%20Server) the constants can be set using a configuration file and read/loaded at runtime. This eliminates the need of the compile and upload workflow. The configuration file can be as simple as

```yaml
left_scale: 0.15
right_scale: 0.15
```

and can even be set in a launch file.

With the addition of `dynamic_reconfigure` the tuning process can be made even easier. The constants can be changed on the fly while the differential drive is running. Now, this will require more work upfront, but the benefits make this additional work well worth it.

I've done some research and found some resources to help with using parameters and `dynamic_reconfigure` with the Arduino.

- [Arduino Motor Control and ROS](http://www.aass.oru.se/Research/Learning/courses/sens/2015/lab1.pdf)
- [associated repo for the above lab](https://github.com/tstoyanov/sensors_lab1/tree/master/arduino_pkg)
- [Using Parameters in roscpp on the ROS wiki](http://wiki.ros.org/roscpp_tutorials/Tutorials/Parameters)
- [Setting up Dynamic Reconfigure for a Node](http://wiki.ros.org/dynamic_reconfigure/Tutorials/SettingUpDynamicReconfigureForANode%28cpp%29)

**Comment by jreyes81** [comment]
Hey josh,

Ive recently been working on this just to get a bit of experience using dynamic reconfiguration. So far I have been able to create a dynamic reconfigurable node that contains the parameters I want to update. There are two ways that I have thought about implementing this reconfiguration in the arduino code to have the parameters update the velocity scales for the wheels of the robot. The first way, which is what I am currently trying and having issues with, is subscribing to the node that is the server node for the dynamic reconfiguration since it publishes out the information that I am changing on-the-go. My issue is being able to correctly access these values , but I feel like I am not doing it correctly or missing something.

The link below is the repository where I have my code:
[https://github.com/asme-calstatela/teleop/tree/jeo](url)

- I have my reconfigurable file that I made in the "cfg" folder
- My created server is called "diff_drive_server.cpp" which is inside the 'src' folder
- The arduino code I am using is called "differential_drive_ver2"

To see how the dynamic reconfiguration works on its own, you can copy the repository since it contains all the necessary dependencies and executables. Then do the following:

1. Run `roscore`
2. Run 'rosrun teleop diff_drive_server'
3. Run 'rosrun rqt_reconfigure rqt_reconfigure'

On the arduino side, the following lines of code deal with what I am trying to do
https://github.com/asme-calstatela/teleop/blob/a86caa5bdbb0585782c8c5bc604a77986f426513/arduino/differential_drive_ver2/differential_drive_ver2.ino#L28
https://github.com/asme-calstatela/teleop/blob/a86caa5bdbb0585782c8c5bc604a77986f426513/arduino/differential_drive_ver2/differential_drive_ver2.ino#L75
https://github.com/asme-calstatela/teleop/blob/9992d43c38af3281e87cfdaa8db72914a50a59c6/arduino/differential_drive_ver2/differential_drive_ver2.ino#L76
https://github.com/asme-calstatela/teleop/blob/eab247f43b1aba76c7ba281948d6eb113f93f02d/arduino/differential_drive_ver2/differential_drive_ver2.ino#L77
https://github.com/asme-calstatela/teleop/blob/eab247f43b1aba76c7ba281948d6eb113f93f02d/arduino/differential_drive_ver2/differential_drive_ver2.ino#L144
https://github.com/asme-calstatela/teleop/blob/eab247f43b1aba76c7ba281948d6eb113f93f02d/arduino/differential_drive_ver2/differential_drive_ver2.ino#L153

When i try to compile it, I get an issue with:
request for member 'value' in 'scale_val.dynamic_reconfigure::Config::doubles', which is of pointer type 'dynamic_reconfigure::DoubleParameter* const' (maybe you meant to use '->' ?)

I have tried looking at how the information is being stored in this topic, but I am getting a bit confused on how to append it. 

The other way I have though about using the dynamic reconfiguration, is through servers and clients in which one of the links you shared on this post, there is a lab that does but I wanted to see if I could do it through a subscribing method first. Thanks for the help

**Comment by jreyes81** [comment]
I just noticed that the links to the specific lines of code make it tedious to look at, so here is how they look:
`#include <dynamic_reconfigure/Config.h> // Updated values for velocity scaling`
`void vel_scale_cb(const dynamic_reconfigure::Config& scale_val){`
` float LEFT_SCALE = scale_val.doubles.value; // Check to see if this is correct`
` float RIGHT_SCALE = scale_val.doubles_length; // Check to see if this correct`
`ros::Subscriber<dynamic_reconfigure::Config> sub_vel_scale("diff_drive_server_node/parameter_updates", vel_scale_cb); //Contains updated velocity scales`
` nh.subscribe(sub_vel_scale); // Dyanmically reconfigured velcity values`

**Comment by jay3ss** [comment]
Hey Jeo,

Can you please post the entire error? I'm not seeing a file name or line number which makes it a little difficult to look into.

Thanks

**Comment by jreyes81** [comment]
Yeah of course. Ill post it when I get back to school on monday if that's fine.

**Comment by jay3ss** [comment]
Yeah, no problem.

**Comment by jay3ss** [comment]
I see some things that are different from the [dyanmic_reconfigure tutorial](http://wiki.ros.org/dynamic_reconfigure/Tutorials/SettingUpDynamicReconfigureForANode%28cpp%29). In the tutorial, the callback function for the `dynamic_reconfigure` server node is

```cpp
void callback(dynamic_tutorials::TutorialsConfig &config, uint32_t level) {
  ROS_INFO("Reconfigure Request: %d %f %s %s %d", 
            config.int_param, config.double_param, 
            config.str_param.c_str(), 
            config.bool_param?"True":"False", 
            config.size);
}
```

Notice that there are _two_ parameters while the callback function in your case has only one

```cpp
void vel_scale_cb(const dynamic_reconfigure::Config& scale_val){
   float LEFT_SCALE = scale_val.doubles.value; // Check to see if this is correct
   float RIGHT_SCALE = scale_val.doubles_length; // Check to see if this correct
}
```

Also, note the way that the tutorial accesses the `double` parameter

```cpp
config.double_param
``` 
versus the way that you're doing it

```cpp
scale_val.doubles.value
```

Notice that you're trying to access some `value` attribute while the tutorial's callback isn't. This is probably the problem because of the error that you're getting (which mentions `value`). Try removing `value` and adding `uint32_t level` as a parameter and see what happens.

**Comment by jay3ss** [comment]
One more thing, you don't need the `float` in the callback either since both `LEFT_SCALE` and `RIGHT_SCALE` are global. So, change

```cpp
float LEFT_SCALE = scale_val.doubles.value;
float RIGHT_SCALE = scale_val.doubles_length;
```

to

```cpp
LEFT_SCALE = scale_val.doubles.value;
RIGHT_SCALE = scale_val.doubles_length;
```

I'm not sure if they should be `double`s instead of `float`s since you have defined them as `double`s in [your `cfg` file](https://github.com/asme-calstatela/teleop/blob/jeo/cfg/diff_drive_con.cfg#L24).
